### PR TITLE
Fix integer overflows when dealing with templates.

### LINKED ISSUE #21622: Template with extreme negative order (e.g. -2147483648) overrides positive order template
**Elasticsearch version**: 5.0.1

**Plugins installed**: []

**JVM version**: 1.8.0_45 but I've seen other versions, too

**OS version**: OSX El Capitan but I've seen it on RHEL 6, too

**Description of the problem including expected versus actual behavior**:
I'm expecting templates with higher `order` to always override those with lower `order`. It doesn't seem to happen if one of them has a value of -2147483648 or close to it (tried with 46 and 47) **AND** another has a positive value (tried here with 2). It works if the negative number is less extreme (-2000000000 for example) **OR** if the second template has a higher, yet still negative `order` (-2 for example).

**Steps to reproduce**:
```
# curl -XDELETE localhost:9200/_template/*
# curl -XDELETE localhost:9200/_all

curl -XPUT localhost:9200/_template/higher_order -d '{
    "template": "*",
    "order": 2,
    "mappings": {
      "message": {
        "properties": {
          "foo": {
            "type": "keyword"
          }
        }
      }
    }
}'

curl -XPUT localhost:9200/_template/lower_order -d '{
    "order" : -2147483648,
    "template" : "*",
    "mappings" : {
      "message" : {
        "properties" : {
          "foo" : {
            "type" : "text"
          }
        }
      }
    }
  }'

curl -XPUT localhost:9200/test
curl localhost:9200/test?pretty
#          "foo" : {
#            "type" : "text"   # <----- should be keyword
#          }
```

This affects Graylog users, see http://docs.graylog.org/en/2.1/pages/configuration/elasticsearch.html#custom-index-mappings

### PR DESCRIPTION
**Pull request by jpountz**
The overflows were happening in two places, the parsing of the template that
implicitly truncates the `order` when its value does not fall into the `integer`
range, and the comparator that sorts templates in ascending order, since it
returns `order2-order1`, which might overflow.

Closes #21622

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/common/Numbers.java [hunk 1]
```diff
@@ -178,4 +181,54 @@ public static boolean isValidDouble(double value) {
...
         return true;
     }
+
+    /** Return the long that {@code n} stores, or throws an exception if the
+     *  stored value cannot be converted to a long that stores the exact same
+     *  value. */
+    public static long toLongExact(Number n) {
+        if (n instanceof Byte || n instanceof Short || n instanceof Integer
+                || n instanceof Long || n instanceof Float || n instanceof Double) {
+            double d = n.doubleValue();
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/Numbers.java]: I think it'd be a bit cleaner not to do this for Byte, Short, Integer, and Long.

**Review by nik9000 — APPROVED** [review]
It'd be nice to have yaml test with a template with a big negative value just to confirm 100% that the fix flows all the way up properly.

**Comment by jpountz** [comment]
> It'd be nice to have yaml test with a template with a big negative value just to confirm 100% that the fix flows all the way up properly.

I wanted to avoid going that route since I don't think we want to have an overflow REST test for every numeric parameter we have in our APIs. So I focused on low-level tests and make sure that this fixed the reported issue manually.

**Comment by nik9000** [comment]
> I don't think we want to have an overflow REST test for every numeric parameter we have in our APIs.

We don't, but since this overflow came up in this context I think it is worth testing it in this context. Others aren't needed.

**Comment by jpountz** [comment]
Sorry but I feel quite strongly against writing a REST test to check for overflows. I'll go with the majority if more people think there should be such a test, but on my end I would like to avoid having such a test.

**Comment by nik9000** [comment]
> Sorry but I feel quite strongly against writing a REST test to check for overflows.

Just merge it then.

**Pull request merged**